### ISSUE TITLE
Replay buffer is storing multiple copies of the most recent episode

### ISSUE THREAD

**Issue by mvindiola1** [Itomigna2/Muesli-lunarlander#1]
Hi @Itomigna2,

This is very cool work. Thanks for sharing it. 

I was testing out your implementation here and noticed a bug. For every new episode you add an entry in the replay buffers for state, action, policy, and reward. But because you are using the same same list over and over you do not get unique episodes in the replay buffer, instead you get references to the same list so they all end up being the exact same most recent episode.

You can check this with this by inserting the code below. Here we are summing the first episode in the replay buffer and then testing if that sum is exactly the same for all of the other episodes in the buffer. You will see that they are.

```
for i in range(episode_nums):
    #...
    print("Size of replay buffer: ", len(agent.action_replay))
    print("All actions same in replay buffer?: ", all([sum(agent.action_replay[0]) == sum(r) for r in agent.action_replay]))
    print("All rewards same in replay buffer? ", all([sum(agent.r_replay[0]) == sum(r) for r in agent.r_replay]))
    agent.update_weights_mu(target)
```

The fix for this is pretty straightforward. We just need to change all of the list.clear() to new lists. Currently you have this at the end of update_weights_mu. I would suggest moving them to the beginning of self_play_mu.
```
        # remove these 
        # self.state_traj.clear()
        # self.action_traj.clear()
        # self.P_traj.clear()
        # self.r_traj.clear()

        self.state_traj = []
        self.action_traj = []
        self.P_traj = []
        self.r_traj = []
```

**Comment by Itomigna2** [comment]
Hi mvindiola1.

I'm very grateful for your help. I checked the code following your suggestion, 
and I found you are exactly right. My code just have worked with on-policy setting. 

I am trying to make progress with changed code in new branch 'testing_issue'.

Some changes (your suggestion, batch size 16 -> 64, online data 25% and replay data 75%, lr 0.00032->0.0003) applied to that code.

The training results,

![11](https://user-images.githubusercontent.com/119741210/217843949-1325ecff-3fd6-4d63-91ea-29251260b15a.png)

https://colab.research.google.com/drive/164h_3ytA_BWOn_0wG8glCHdub4oXJ-Yp?usp=sharing

It learns faster, but has high variance. (Probably due to the absence of retrace estimator)



Thank you again for your perfect suggestion. It was very helpful!


Best regards,

Itomigna2
